core/lib/dal/src/consensus_dal/mod.rs

```diff
@@ -518,25 +538,33 @@ impl ConsensusDal<'_, '_> {
     /// Fails if certificate doesn't match the stored block.
     pub async fn insert_block_certificate(
         &mut self,
-        cert: &validator::CommitQC,
+        cert: &BlockCertificate,
     ) -> Result<(), InsertCertificateError> {
         use InsertCertificateError as E;
-        let header = &cert.message.proposal;
+
+        // Extract block number and payload hash based on certificate variant
+        let (block_number, payload_hash) = match cert {
+            BlockCertificate::V1(qc) => (qc.message.proposal.number, qc.message.proposal.payload),
+            BlockCertificate::V2(qc) => (qc.message.proposal.number, qc.message.proposal.payload),
+        };
```

Comment:
Why for some things we have methods in BlockCertificate like (like `number`), whereas for other we `match` on the caller side (like here or above in `block_store_state`)? 

I think it's a good idea to encapsulate versioning logic in the  enum itself, so that you don't have to modify it everywhere when you add a new version. Also it's easier to follow the caller logic (you have to deal with less concepts there)

Reply:
No particular reason. I'll encapsulate all the methods then.